File: app/src/main/java/com/liweiyap/narradir/ui/ObserverImageButton.java
package com.liweiyap.narradir.ui;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.liweiyap.narradir.util.IObserver;

import java.util.ArrayList;
import java.util.List;

public class ObserverImageButton
    extends androidx.appcompat.widget.AppCompatImageButton
    implements View.OnClickListener, View.OnLongClickListener, IObserverListener
{
    public ObserverImageButton(@NonNull Context context)
    {
        super(context);
        setOnClickListener(this);
        setOnLongClickListener(this);
    }

    public ObserverImageButton(@NonNull Context context, @Nullable AttributeSet attrs)
    {
        super(context, attrs);
        setOnClickListener(this);
        setOnLongClickListener(this);
    }

    public ObserverImageButton(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr)
    {
        super(context, attrs, defStyleAttr);
        setOnClickListener(this);
        setOnLongClickListener(this);
    }

    public void destroy()
    {
        clearOnClickObservers();
        clearOnLongClickObservers();

        mOnClickObservers = null;
        mOnLongClickObservers = null;

        setOnClickListener(null);
        setOnLongClickListener(null);
    }

    @Override
    public void onClick(View view)
    {
        notifyOnClickObservers();
    }

    @Override
    public boolean onLongClick(View view)
    {
        notifyOnLongClickObservers();
        return true;  // https://stackoverflow.com/a/3756619/12367873; https://stackoverflow.com/questions/4402740/android-long-click-on-a-button-perform-actions
    }

    @Override
    public void addOnClickObserver(final IObserver observer)
    {
        if (mOnClickObservers == null)
        {
            return;
        }

        mOnClickObservers.add(observer);
    }

    @Override
    public void notifyOnClickObservers()
    {
        if (mOnClickObservers == null)
        {
            return;
        }

        mOnClickObservers.forEach(IObserver::update);
    }

    @Override
    public void clearOnClickObservers()
    {
        if (mOnClickObservers == null)
        {
            return;
        }

        mOnClickObservers.clear();
    }

    @Override
    public void addOnLongClickObserver(final IObserver observer)
    {
        if (mOnLongClickObservers == null)
        {
            return;
        }

        mOnLongClickObservers.add(observer);
    }

    @Override
    public void notifyOnLongClickObservers()
    {
        if (mOnLongClickObservers == null)
        {
            return;
        }

        mOnLongClickObservers.forEach(IObserver::update);
    }

    @Override
    public void clearOnLongClickObservers()
    {
        if (mOnLongClickObservers == null)
        {
            return;
        }

        mOnLongClickObservers.clear();
    }

    protected List<IObserver> mOnClickObservers = new ArrayList<>();
    protected List<IObserver> mOnLongClickObservers = new ArrayList<>();
}

File: app/src/main/java/com/liweiyap/narradir/ui/ObserverImageButton.kt
package com.liweiyap.narradir.ui

import android.content.Context
import android.util.AttributeSet
import android.view.View

import androidx.appcompat.widget.AppCompatImageButton

open class ObserverImageButton:
    AppCompatImageButton,
    View.OnClickListener, View.OnLongClickListener, IObserverListener
{
    constructor(context: Context): super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?): super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int): super(context, attrs, defStyleAttr) {
        init()
    }

    private fun init() {
        setOnClickListener(this)
        setOnLongClickListener(this)
    }

    fun destroy() {
        clearOnClickObservers()
        clearOnLongClickObservers()

        mOnClickObservers = null
        mOnLongClickObservers = null

        setOnClickListener(null)
        setOnLongClickListener(null)
    }

    override fun onClick(view: View) {
        notifyOnClickObservers()
    }

    override fun onLongClick(view: View): Boolean {
        notifyOnLongClickObservers()
        return true  // https://stackoverflow.com/a/3756619/12367873; https://stackoverflow.com/questions/4402740/android-long-click-on-a-button-perform-actions
    }

    override fun addOnClickObserver(observer: () -> Unit) {
        mOnClickObservers?.add(observer)
    }

    override fun notifyOnClickObservers() {
        mOnClickObservers?.forEach { observer: () -> Unit ->
            observer()
        }
    }

    override fun clearOnClickObservers() {
        mOnClickObservers?.clear()
    }

    override fun addOnLongClickObserver(observer: () -> Unit) {
        mOnLongClickObservers?.add(observer)
    }

    override fun notifyOnLongClickObservers() {
        mOnLongClickObservers?.forEach { observer: () -> Unit ->
            observer()
        }
    }

    override fun clearOnLongClickObservers() {
        mOnLongClickObservers?.clear()
    }

    private var mOnClickObservers: MutableList<() -> Unit>? = ArrayList()
    private var mOnLongClickObservers: MutableList<() -> Unit>? = ArrayList()
}